### PR TITLE
Ultra

### PR DESCRIPTION
**Pull request by benjamin-wilson**
Pull 1366 work into main. This branch supports both 1366 and 1397 with configuration in menuconfig.

### CODE REVIEW

File: components/bm1397/include/bm1366.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#ifndef BM1366_H_
+#define BM1366_H_
+
+#include "driver/gpio.h"
+#include "bm1397.h"
+
+
+#define CRC5_MASK 0x1F
+
+static const u_int64_t BM1366_FREQUENCY = CONFIG_BM1397_FREQUENCY;
```
Comment (Georges760) [inline components/bm1397/include/bm1366.h:10]: if this CONFIG is common, shouldn't we call it `CONFIG_ASIC_FREQUENCY` ?

File: main/nvs_config.h [hunk 2]
```diff
@@ -11,6 +11,7 @@
 #define NVS_CONFIG_STRATUM_PASS "stratumpass"
 #define NVS_CONFIG_BM1397_FREQ "bm1397frequency"
```
Comment (Georges760) [inline main/nvs_config.h:12]: since we are multi-asic now, we should rename BM1397 into ASIC into this file

File: components/bm1397/bm1366.c [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+}
+
+static void _set_chip_address(uint8_t chipAddr) {
+
+    unsigned char read_address[2] = {chipAddr, 0x00};
+    //send serial data
+    _send_BM1366((TYPE_CMD | GROUP_SINGLE | CMD_SETADDRESS), read_address, 2, false);
+}
+
+static unsigned char _reverse_bits(unsigned char num) {
```
Comment (Georges760) [inline components/bm1397/bm1366.c:127]: look like a lot of helper function can be factorized, like this one. Can you created a common.c/.h in the bm1397 component ?

**Review by Georges760 — CHANGES_REQUESTED** [review]

File: components/stratum/stratum_api.c [hunk 4]
```diff
@@ -254,7 +254,7 @@ int STRATUM_V1_subscribe(int socket, char ** extranonce, int * extranonce2_len)
 {
     // Subscribe
     char subscribe_msg[BUFFER_SIZE];
-    sprintf(subscribe_msg, "{\"id\": %d, \"method\": \"mining.subscribe\", \"params\": [\"bitaxe v2.2\"]}\n", send_uid++);
+    sprintf(subscribe_msg, "{\"id\": %d, \"method\": \"mining.subscribe\", \"params\": [\"bitaxe 1366 V1\"]}\n", send_uid++);
```
[the remark this answers is not in the document]
Comment (johnny9) [reply components/stratum/stratum_api.c:257]: Either a config or add a parameter. My preference would be a parameter.

File: components/bm1397/bm1397.c [hunk 5]
```diff
@@ -301,10 +338,127 @@ void BM1397_set_job_difficulty_mask(int difficulty){
...
+
+
+    GlobalState *GLOBAL_STATE = (GlobalState*) pvParameters;
+    if (GLOBAL_STATE->valid_jobs[rx_job_id] == 0) {
+        ESP_LOGI(TAG, "Invalid job nonce found, id=%d", rx_job_id);
+    }
+
+    uint32_t rolled_version = GLOBAL_STATE->ASIC_TASK_MODULE.active_jobs[rx_job_id]->version;
+    for (int i = 0; i < rx_midstate_index; i++) {
+        rolled_version = increment_bitmask(rolled_version, GLOBAL_STATE->ASIC_TASK_MODULE.active_jobs[rx_job_id]->version_mask);
```
[the remark this answers is not in the document]
Comment (johnny9) [reply components/bm1397/bm1397.c:436]: We roll the version for the midstates for the 1397

File: components/bm1397/CMakeLists.txt [hunk 6]
```diff
@@ -9,4 +10,11 @@ REQUIRES
     freertos 
     driver 
     stratum
-)
\ No newline at end of file
+)
+
+
+# Include the header files from "main" directory
+target_include_directories(${COMPONENT_LIB} PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/../../main")
```
Comment (johnny9) [inline components/bm1397/CMakeLists.txt:17]: The bm13xx component shouldn't depend on our main component

File: components/bm1397/bm1397.c [hunk 7]
```diff
@@ -301,10 +338,127 @@ void BM1397_set_job_difficulty_mask(int difficulty){
...
+    }
+
+    if (asic_result->nonce == prev_nonce) {
+        return NULL;
+    } else {
+        prev_nonce = asic_result->nonce;
+    }
+
+
+    result.job_id = rx_job_id;
```
[the remark this answers is not in the document]
Comment (benjamin-wilson) [reply components/bm1397/bm1397.c:434]: There is an asic job id and stratum job id. The stratum job id is fetched in asic_result_task.c. They should probably be renamed but for the purposes of merging I'll wait.

**Comment by benjamin-wilson** [comment]
Resolved suggestions

File: main/global_state.h [hunk 8]
```diff
@@ -3,16 +3,28 @@
...
+typedef struct {
+    void (*init_fn)(u_int64_t);
+    task_result * (*receive_result_fn)(void *GLOBAL_STATE);
+    int (*set_max_baud_fn)(void);
+    void (*set_difficulty_mask_fn)(int);
+    void (*send_work_fn)(void *GLOBAL_STATE, bm_job *next_bm_job);
+} AsicFunctions;
 
 typedef struct  {
+    AsicFunctions ASIC_FUNCTIONS;
```
Comment (johnny9) [inline main/global_state.h:25]: properties shouldnt be all caps. makes it look like a #define

File: components/bm1397/bm1397.c [hunk 9]
```diff
@@ -12,6 +12,28 @@
 #include "bm1397.h"
 #include "utils.h"
 #include "crc.h"
+#include "mining.h"
+#include "global_state.h"
```
Comment (johnny9) [inline components/bm1397/bm1397.c:16]: bm1397 shouldnt have to import global_state.h

**Review by johnny9 — COMMENTED** [review]
Please squash commit history

File: components/bm1397/bm1397.c (hunk 9 above)
Comment (benjamin-wilson) [reply components/bm1397/bm1397.c:16]: It needs to refrence the global state struct for BM1397_send_work etc.

Comment (johncarl-te) [reply components/bm1397/bm1397.c:16]: Yeah that likely means there is something wrong with the api then.

**Pull request merged**